modules/extension/transform/src/main/java/org/geotools/data/transform/TransformFeatureCollection.java

```diff
@@ -206,4 +216,56 @@ public SimpleFeatureCollection subCollection(Filter filter) {
         }
         return new TransformFeatureCollection(source, transformer, q);
     }
+
+    /**
+     * Checks if the visitor is accessing only properties available in the specified feature type,
+     * checks if the target schema contains transformed attributes or as a special case, if it's a
+     * count visitor accessing no properties at all
+     *
+     * @param visitor
+     * @param featureType
+     * @return
+     */
+    protected boolean isTypeCompatible(FeatureVisitor visitor, SimpleFeatureType featureType) {
```

Comment:
Verified better and these tests are not valid, could have seen this easily by also checking the results of the visits in tests, like I just did. Since it's been a while (sorry, summer vacations with family and conferences kept all my spare time occupied) I'm going to fix it directly.